code/hsec-sync/test/Spec/SyncSpec.hs

```diff
@@ -8,79 +7,87 @@ import Data.Bifunctor (first)
           testCase "Sync behind should update" $
             withSystemTempDirectory "hsec-sync" $ \p -> do
-              let repo = withRepositoryAt p
-              resultCreate <- sync repo
+              let snapshot = withSnapshotAt p
+              resultCreate <- sync snapshot
+              resultCreate @?= Right Created
+              writeFile
+                (p </> "snapshot.json")
+                "{\"latestUpdate\":\"2020-03-11T12:26:51Z\",\"snapshotVersion\":\"0.1.0.0\"}"
+              status snapshot >>= (@?= DirectoryOutDated)
+              resultResync <- sync snapshot
+              resultResync @?= Right Updated
+              status snapshot >>= (@?= DirectoryUpToDate),
+          testCase "Sync a broken snapshot.json" $
+            withSystemTempDirectory "hsec-sync" $ \p -> do
+              let snapshot = withSnapshotAt p
+              resultCreate <- sync snapshot
               resultCreate @?= Right Created
-              D.withCurrentDirectory p $ do
-                (statusReset, _, _) <-
-                  readProcessWithExitCode "git" ["reset", "--hard", "HEAD~50"] ""
-                statusReset @?= ExitSuccess
-              status repo >>= (@?= DirectoryOutDated)
-              resultResync <- sync repo
+              writeFile
+                (p </> "snapshot.json")
+                "{\"latestpdate\":\"2020-03-11T12:26:51Z\",\"snapshotVersion\":\"0.1.0.0\"}"
+              status snapshot >>= (@?= DirectoryIncoherent)
+              resultResync <- sync snapshot
               resultResync @?= Right Updated
-              status repo >>= (@?= DirectoryUpToDate),
-          testCase "Sync behind and changed remote should update" $
+              status snapshot >>= (@?= DirectoryUpToDate),
+          testCase "Sync a deleted snapshot.json" $
             withSystemTempDirectory "hsec-sync" $ \p -> do
-              let repo = withRepositoryAt p
-              resultCreate <- sync repo
+              let snapshot = withSnapshotAt p
+              resultCreate <- sync snapshot
               resultCreate @?= Right Created
-              D.withCurrentDirectory p $ do
-                (statusReset, _, _) <-
-                  readProcessWithExitCode "git" ["reset", "--hard", "HEAD~50"] ""
-                statusReset @?= ExitSuccess
-                (statusRemote, _, _) <-
-                  readProcessWithExitCode "git" ["remote", "rename", "origin", "old"] ""
-                statusRemote @?= ExitSuccess
-              resultResync <- sync repo
+              D.removeFile (p </> "snapshot.json")
+              status snapshot >>= (@?= DirectoryOutDated)
+              resultResync <- sync snapshot
               resultResync @?= Right Updated
+              status snapshot >>= (@?= DirectoryIncoherent)
         ]
     ]
 
-withRepositoryAt :: FilePath -> Repository
-withRepositoryAt root =
-  defaultRepository {repositoryRoot = root}
+withSnapshotAt :: FilePath -> Snapshot
+withSnapshotAt root =
```

Comment:
the naming here might be a bit unfortunate; `withSomething` usually denotes a function that is of type `Codensity m Something` or `_ -> Codensity m Something`